packages/vscode-extension/src/treeview/environmentTreeViewProvider.ts

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { Mutex } from "async-mutex";
+import * as vscode from "vscode";
+
+import { err, FxError, LocalEnvironmentName, ok, Result, Void } from "@microsoft/teamsfx-api";
+import { environmentManager } from "@microsoft/teamsfx-core";
+
+import { ext } from "../extensionVariables";
+import { DynamicNode } from "./dynamicNode";
+import { EnvironmentNode } from "./environmentTreeItem";
+
+export class EnvironmentTreeViewProvider implements vscode.TreeDataProvider<DynamicNode> {
+  private static instance: EnvironmentTreeViewProvider;
+  private _onDidChangeTreeData: vscode.EventEmitter<DynamicNode | undefined | void> =
+    new vscode.EventEmitter<DynamicNode | undefined | void>();
+  readonly onDidChangeTreeData: vscode.Event<DynamicNode | undefined | void> =
+    this._onDidChangeTreeData.event;
+
+  private environments: DynamicNode[] = [];
+  private mutex = new Mutex();
+
+  private constructor() {}
+
+  public static getInstance() {
+    if (!EnvironmentTreeViewProvider.instance) {
+      EnvironmentTreeViewProvider.instance = new EnvironmentTreeViewProvider();
+    }
+    return EnvironmentTreeViewProvider.instance;
+  }
+
+  public async reloadEnvironments(): Promise<Result<Void, FxError>> {
+    if (!ext.workspaceUri) {
```

Comment:
Maybe check whether it's teamsfx project here so we don't need to continue for non-teamsfx projects?

Reply:
Thanks, fixed.